src/riak_cs_gc.erl

```diff
@@ -181,26 +181,41 @@ gc_specific_manifests(UUIDsToMark, RiakObject, Bucket, Key, RcPid) ->
     {error, term()} | {ok, riakc_obj:riakc_obj()}.
 handle_mark_as_pending_delete({ok, RiakObject}, Bucket, Key, UUIDsToMark, RcPid) ->
     Manifests = riak_cs_manifest:manifests_from_riak_object(RiakObject),
-    PDManifests = riak_cs_manifest_utils:manifests_to_gc(UUIDsToMark, Manifests),
-    MoveResult = move_manifests_to_gc_bucket(PDManifests, RcPid),
+    PDManifests0 = riak_cs_manifest_utils:manifests_to_gc(UUIDsToMark, Manifests),
+    {PDManifests, UUIDsToDelete} =
+        case riak_cs_config:active_delete_threshold() of
+            Threshold when is_integer(Threshold) andalso Threshold > 0 ->
+                %% hereby delete PDUUIDs smaller than threshold
+                
+                %% We do synchronous delete after it is marked
+                %% pending_delete, to reduce the possibility where
+                %% concurrent requests find active manifest (UUID) and
+                %% go find deleted blocks resulting notfound stuff.
+                %% However, there are still corner cases where
+                %% concurrent requests interleaves between marking
+                %% pending_delete here and deleting blocks, like:
+                %%
+                %% 1. Request A refers to a manifest finding active UUID x
+                %% 2. Request B deletes an object marking active UUID x as pending_delete
+                %% 3. Request B deletes blocks of UUID x according to this synchronous delete -> ok
+                %% 4. Request A refers to blocks pointed by UUID x -> notfound
+                maybe_delete_small_objects(PDManifests0, RcPid, Threshold);
+            _ ->
+                {PDManifests0, []}
+        end,
+
     PDUUIDs = [UUID || {UUID, _} <- PDManifests],
-    handle_move_result(MoveResult, RiakObject, Bucket, Key, PDUUIDs, RcPid);
+    case move_manifests_to_gc_bucket(PDManifests, RcPid) of
+        ok ->
+            mark_as_scheduled_delete(PDUUIDs, UUIDsToDelete, RiakObject, Bucket, Key, RcPid);
```

Comment:
Currently, `riak_cs_gc:mark_manifests()` attempts to update manifest objects even `UUIDsToMark` is empty list.
As possibility of empty list grows with this PR, adding a clause to avoid update for empty `UUIDsToMark` list would be preferred.